doc/manual/src/language/import-from-derivation.md

```diff
@@ -0,0 +1,87 @@
+# Import From Derivation
+
+The value of a Nix expression can depend on the contents of a [store object].
+In this case, when that store object is needed, evaluation will be paused, the store object [realised], and then evaluation resumed.
+
+[store object]: @docroot@/glossary.md#gloss-store-object
+[derivation]: @docroot@/glossary.md#gloss-derivation
+[realised]: @docroot@/glossary.md#gloss-realise
+
+This has performance implications:
+Since evaluation is sequential, each required store object that is not already in the store will also be realised sequentially.
+
+Passing a store path to any built-in function that reads from the filesystem constitutes Import From Derivation:
+
+- [`import`](./builtins.md#builtins-import)` expr`
+- [`builtins.readFile`](./builtins.md#builtins-readFile)` expr`
+- [`builtins.readFileType`](./builtins.md#builtins-readFileType)` expr`
+- [`builtins.readDir`](./builtins.md#builtins-readDir)` expr`
+- [`builtins.pathExists`](./builtins.md#builtins-pathExists)` expr`
+- [`builtins.filterSource`](./builtins.md#builtins-filterSource)` f expr`
+- [`builtins.path`](./builtins.md#builtins-path)` { path = expr; }`
+- [`builtins.hashFile`](./builtins.md#builtins-hashFile)` t expr`
+- `builtins.scopedImport x drv`
+
+Realising store objects during evaluation can be disabled by setting [`allow-import-from-derivation`](../command-ref/conf-file.md#conf-allow-import-from-derivation) to `false`.
+
+## Example
+
+In the following Nix expression, the inner derivation `drv` produces a file containing `"hello"`.
+
+```nix
+# IFD.nix
+let
+  drv = derivation {
+    name = "hello";
+    builder = "/bin/sh";
+    args = [ "-c" ''echo \"hello\" > $out'' ];
+    system = builtins.currentSystem;
+  };
+in "${import drv} world"
+```
```

Comment:
````suggestion
In the following Nix expression, the inner derivation `drv` produces a file with contents `hello`.

```nix
# IFD.nix
let
  drv = derivation {
    name = "hello";
    builder = "/bin/sh";
    args = [ "-c" "echo hello > $out" ];
    system = builtins.currentSystem;
  };
in "${readFile drv} world"
```
````

This makes for slightly less quotation noise.

Reply:
This adds a newline after `hello`, because of `echo`.
Also I think we should use the IFD in a subsequent derivation to prove a point later with the `nix-build` log.

How about this:

````suggestion
In the following Nix expression, the inner derivation `drv` produces a file with contents `hello`.

# IFD.nix
let
  drv = derivation {
    name = "hello";
    builder = "/bin/sh";
    args = [ "-c" "echo -n hello > $out" ];
    system = builtins.currentSystem;
  };
  # import from derivation
  greeting = builtins.readFile drv;
in

derivation {
  name = builtins.replaceStrings ["h"] ["H"] greeting;
  builder = "/bin/sh";
  args = [ "-c" "echo ${greeting} > $out" ];
  system = builtins.currentSystem;
}
```
````

Reply:
````suggestion
In the following Nix expression, the inner derivation `drv` produces a file with contents `hello`.

```nix
# IFD.nix
let
  drv = derivation {
    name = "hello";
    builder = "/bin/sh";
    args = [ "-c" "echo -n hello > $out" ];
    system = builtins.currentSystem;
  };
in "${builtins.readFile drv} world"
```
````

Add `echo -n` to make it work as expected.

Reply:
> Not sure if it's still as snappy as `import drv`

Not using `import` is in line with the point that it's not just about `import`, but it might also confuse, as it's not quite as clear where the IFD happens.

````suggestion
In the following Nix expression, the inner derivation `drv` produces a file containing `"hello"`.

```nix
# IFD.nix
let
  drv = derivation {
    name = "hello";
    builder = "/bin/sh";
    args = [ "-c" "echo -n hello > $out" ];
    system = builtins.currentSystem;
  };
  # evaluating contents requires `drv` to be realised,
  # also known as import from derivation
  contents = builtins.readFile drv;
in "${contents} world"
```
````